### PR TITLE
Adding the Aggregate State Controller

### PR DESCRIPTION
**Pull request by bradmwilliams**
Refactoring JobStatus methods

Adding JobRunResult helpers and test

Adding aggregate_state_controller logic

Refactoring status to releasepayloadstatus

Enabling the Aggregate State Controller

### CODE REVIEW

File: pkg/cmd/release-payload-controller/aggregate_state_controller.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	originalReleasePayload, err := c.releasePayloadLister.ReleasePayloads(namespace).Get(name)
+	// The ReleasePayload resource may no longer exist, in which case we stop processing.
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	releasePayload := originalReleasePayload.DeepCopy()
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/aggregate_state_controller.go:83]: at larger scales, you will need to figure out how to know if you need to change a value *before* the deepcopy.

File: pkg/releasepayload/jobrunresult/helpers.go [hunk 2]
```diff
@@ -16,3 +19,129 @@ func (in ByCoordinatesName) Len() int {
...
+	for i := range results {
+		if cmp.Equal(results[i].Coordinates, coordinates) {
+			return &results[i]
+		}
+	}
+
+	return nil
+}
+
+func ComputeAggregateState(job v1alpha1.JobStatus) v1alpha1.JobState {
```
Comment (deads2k) [inline pkg/releasepayload/jobrunresult/helpers.go:63]: single call site?  If so, put it in that controller file.

File: pkg/releasepayload/jobrunresult/helpers.go [hunk 3]
```diff
@@ -16,3 +19,129 @@ func (in ByCoordinatesName) Len() int {
...
+			failedJobs = append(failedJobs, result)
+		default:
+			unknownJobs = append(unknownJobs, result)
+		}
+	}
+
+	return pendingJobs, successfulJobs, failedJobs, unknownJobs
+}
+
+func processJobResults(results []v1alpha1.JobRunResult) v1alpha1.JobState {
```
Comment (deads2k) [inline pkg/releasepayload/jobrunresult/helpers.go:98]: this is actually a processSingleJobResult, right?  Without an array?

File: pkg/releasepayload/jobrunresult/helpers.go [hunk 4]
```diff
@@ -16,3 +19,129 @@ func (in ByCoordinatesName) Len() int {
...
+	var pendingJobs, successfulJobs, failedJobs, _ = interrogateJobResults(results)
+
+	switch {
+	case len(successfulJobs) > 0:
+		return v1alpha1.JobStateSuccess
+	case len(pendingJobs) > 0:
+		return v1alpha1.JobStatePending
+	case len(failedJobs) > 0:
```
Comment (deads2k) [inline pkg/releasepayload/jobrunresult/helpers.go:106]: check failures first

File: pkg/releasepayload/jobrunresult/helpers.go (hunk 3 above)
Comment (deads2k) [inline pkg/releasepayload/jobrunresult/helpers.go:98]: you need to doc this.  I didn't know what it did.

File: pkg/releasepayload/jobrunresult/helpers.go [hunk 5]
```diff
@@ -16,3 +19,129 @@ func (in ByCoordinatesName) Len() int {
...
+	var pendingJobs, successfulJobs, failedJobs, _ = interrogateJobResults(results)
+
+	switch {
+	case len(results) != count:
+		return v1alpha1.JobStateUnknown
+	case len(pendingJobs) > 0:
+		return v1alpha1.JobStatePending
+	case len(failedJobs) > 0:
+		return v1alpha1.JobStateFailure
+	case len(successfulJobs) == count:
```
Comment (deads2k) [inline pkg/releasepayload/jobrunresult/helpers.go:123]: let's call a success threshold here 60% success.

File: pkg/releasepayload/jobstatus/helpers.go [hunk 6]
```diff
@@ -19,7 +19,47 @@ func (in ByJobStatusCIConfigurationName) Swap(i, j int) {
 	in[i], in[j] = in[j], in[i]
 }
 
-func ComputeAggregatedJobState(jobs []v1alpha1.JobStatus) v1alpha1.JobState {
+func SetJobStatus(results *[]v1alpha1.JobStatus, newResult v1alpha1.JobStatus) {
```
Comment (deads2k) [inline pkg/releasepayload/jobstatus/helpers.go:22]: if you have one call site, put it in that file.

File: pkg/cmd/release-payload-controller/cmd.go [hunk 7]
```diff
@@ -154,6 +160,7 @@ func (o *Options) Run(ctx context.Context) error {
 	go payloadAcceptedController.RunWorkers(ctx, 1)
 	go payloadRejectedController.RunWorkers(ctx, 1)
 	go pjController.RunWorkers(ctx, 1)
+	go aggregateStateController.RunWorkers(ctx, 1)
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/cmd.go:163]: be brave, try 10

File: pkg/cmd/release-payload-controller/aggregate_state_controller.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+func NewAggregateStateController(
+	releasePayloadNamespace string,
+	releasePayloadInformer releasepayloadinformer.ReleasePayloadInformer,
+	releasePayloadClient releasepayloadclient.ReleaseV1alpha1Interface,
+	eventRecorder events.Recorder,
+) (*AggregateStateController, error) {
+	c := &AggregateStateController{
+		ReleasePayloadController: NewReleasePayloadController("Aggregate State Controller",
```
[the remark this answers is not in the document]
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/aggregate_state_controller.go]: Renamed to `JobStatusAggregateStateController`

File: pkg/releasepayload/jobrunresult/helpers.go (hunk 2 above)
Comment (bradmwilliams) [reply pkg/releasepayload/jobrunresult/helpers.go]: Done

File: pkg/releasepayload/jobrunresult/helpers.go [hunk 9]
```diff
@@ -16,3 +19,129 @@ func (in ByCoordinatesName) Len() int {
...
+func processRetryJobResults(results []v1alpha1.JobRunResult, count int) v1alpha1.JobState {
+	attempts := len(results)
+
+	state := processJobResults(results)
+
+	switch state {
+	case v1alpha1.JobStateSuccess, v1alpha1.JobStatePending:
+		return state
+	case v1alpha1.JobStateFailure:
+		switch {
```
[the remark this answers is not in the document]
Comment (bradmwilliams) [reply pkg/releasepayload/jobrunresult/helpers.go]: Done

File: pkg/cmd/release-payload-controller/cmd.go (hunk 7 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/cmd.go:149]: Done

File: pkg/releasepayload/jobrunresult/helpers.go (hunk 3 above)
Comment (bradmwilliams) [reply pkg/releasepayload/jobrunresult/helpers.go]: Done

File: pkg/cmd/release-payload-controller/aggregate_state_controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/aggregate_state_controller.go]: Renamed (again) to: `JobStateController`

File: pkg/releasepayload/jobrunresult/helpers.go (hunk 5 above)
Comment (bradmwilliams) [reply pkg/releasepayload/jobrunresult/helpers.go]: Done

File: pkg/cmd/release-payload-controller/aggregate_state_controller.go (hunk 1 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/aggregate_state_controller.go]: I've added a `TODO` here and I'll revisit in a follow on PR.

File: pkg/cmd/release-payload-controller/job_state_controller.go [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"reflect"
+
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+const (
+	analysisSuccessThreshold = 0.6
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/job_state_controller.go:23]: requires doc.

Comment (deads2k) [reply pkg/cmd/release-payload-controller/job_state_controller.go:23]: wait, one call site?   move it down near usage

File: pkg/cmd/release-payload-controller/job_state_controller.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+			unknownJobs = append(unknownJobs, result)
+		}
+	}
+
+	return pendingJobs, successfulJobs, failedJobs, unknownJobs
+}
+
+// processJobResults processes JobRunResults for jobs that are not configured as analysis jobs or
+// that have retries.  Any Successful result of this class of jobs returns a "Success" JobState.
+func processJobResults(results []v1alpha1.JobRunResult) v1alpha1.JobState {
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/job_state_controller.go:159]: note to self, this looks fine.

File: pkg/cmd/release-payload-controller/job_state_controller.go [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+// processAnalysisJobResults processes JobRunResults for analysis jobs.  The JobState will be
+// "Unknown" until the number of JobRunResults equals count.  While any of the JobRunResults are
+// still running (pending), the JobState will be "Pending".  Once all the JobRunResults are complete,
+// a success rate, that is greater-than or equal to analysisSuccessThreshold, will return a "Success"
+// JobState.  Anything less-than analysisSuccessThreshold will return a "Failure" JobState.
+func processAnalysisJobResults(results []v1alpha1.JobRunResult, count int) v1alpha1.JobState {
+	var pendingJobs, successfulJobs, _, _ = interrogateJobResults(results)
+
+	switch {
+	case len(results) != count:
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/job_state_controller.go:183]: >=?

File: pkg/cmd/release-payload-controller/job_state_controller.go [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+		return v1alpha1.JobStateUnknown
+	case len(pendingJobs) > 0:
+		return v1alpha1.JobStatePending
+	case (float64(len(successfulJobs)) / float64(count)) >= analysisSuccessThreshold:
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/job_state_controller.go:187]: if we already passed enough, why do we have to wait for pending jobs to finish?

File: pkg/cmd/release-payload-controller/job_state_controller.go (hunk 12 above)
Comment (deads2k) [reply pkg/cmd/release-payload-controller/job_state_controller.go:183]: `< count`

File: pkg/cmd/release-payload-controller/job_state_controller.go [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+// processRetryJobResults processes JobRunResults for jobs that can be retried.  Any successful
+// JobRunResult will return a "Success" JobStatus.  As long as the number of attempts is less than
+// count and each attempt results in a failure, the JobStatus will return "Pending".  If all retry
+// attempts have failed, the JobStatus will be "Failed".
+func processRetryJobResults(results []v1alpha1.JobRunResult, count int) v1alpha1.JobState {
+	attempts := len(results)
+
+	state := processJobResults(results)
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/job_state_controller.go:201]: please rewrite with interrogating here and write the required logic for this method.

**Comment by deads2k** [comment]
minor comments, lgtm otherwise.

File: pkg/cmd/release-payload-controller/job_state_controller.go (hunk 10 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/job_state_controller.go:23]: Done!  (although it pains me :smile: )

File: pkg/cmd/release-payload-controller/job_state_controller.go (hunk 12 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/job_state_controller.go:183]: Done

File: pkg/cmd/release-payload-controller/job_state_controller.go (hunk 13 above)
Comment (bradmwilliams) [reply pkg/cmd/release-payload-controller/job_state_controller.go:187]: Switched order. 
Done

**Comment by deads2k** [comment]
Forgot to push?

File: pkg/cmd/release-payload-controller/job_state_controller.go [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+// JobState.  Anything less-than analysisSuccessThreshold will return a "Failure" JobState.
+func processAnalysisJobResults(results []v1alpha1.JobRunResult, count int) v1alpha1.JobState {
+	var pendingJobs, successfulJobs, _, _ = interrogateJobResults(results)
+
+	switch {
+	case len(results) < count:
+		return v1alpha1.JobStateUnknown
+	case (float64(len(successfulJobs)) / float64(count)) >= analysisSuccessThreshold:
+		return v1alpha1.JobStateSuccess
+	case len(pendingJobs) > 0:
```
Comment (deads2k) [inline pkg/cmd/release-payload-controller/job_state_controller.go:189]: If it mathematically impossible to get enough passes with the remaining jobs you could fail

**Pull request merged**